{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Modules\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from Bio import SeqIO\n",
    "from Bio.Seq import Seq\n",
    "import re\n",
    "import numbers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Functions\n",
    "def read_fasta(name):\n",
    "    fasta_seqs = SeqIO.parse(open(name + '.fa'),'fasta')\n",
    "    data = []\n",
    "    for fasta in fasta_seqs:\n",
    "        data.append([fasta.id, str(fasta.seq).strip().upper()])\n",
    "            \n",
    "    return data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Data\n",
    "grna_list = pd.read_csv('grna_list_to_rank.csv')  \n",
    "gene_list = pd.read_csv('sd108_genetable.csv')  \n",
    "name_conversion = pd.read_csv('protein_mrna_gene_transcript_ids.csv', index_col=0)\n",
    "in_fasta_1 = 'sd108'\n",
    "seqs_df = pd.DataFrame(read_fasta(in_fasta_1), columns = ['name', 'sequence'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Gene Density\n",
    "gene_den = []\n",
    "genes_in_vicinity = []\n",
    "\n",
    "for i in range(np.shape(grna_list)[0]):\n",
    "    count = 0\n",
    "    grna_sca = grna_list['scaffold'][i]\n",
    "    grna_loc = grna_list['loc'][i]\n",
    "    grna_strand = grna_list['strand'][i]\n",
    "    curr_gene_vicinity = []\n",
    "    \n",
    "    curr_goi = gene_list.loc[gene_list['chrom'] == grna_sca].reset_index(drop=True)\n",
    "    if np.shape(curr_goi)[0] > 0:\n",
    "        scaffold_len = len(seqs_df['sequence'][int(grna_sca.split('_')[1])-1])\n",
    "        if grna_strand == '+':\n",
    "            r_limit = grna_loc + 10000\n",
    "            l_limit = grna_loc - 10000   \n",
    "        else:\n",
    "            r_limit = scaffold_len - grna_loc + 10000\n",
    "            l_limit = scaffold_len - grna_loc - 10000\n",
    "    \n",
    "        if r_limit > scaffold_len:\n",
    "            r_limit = scaffold_len\n",
    "        if l_limit < 0:\n",
    "            l_limit = 0\n",
    "    \n",
    "        for j in range(np.shape(curr_goi)[0]):\n",
    "            if curr_goi['txStart'][j] > l_limit and curr_goi['txEnd'][j] < r_limit:\n",
    "                curr_gene_vicinity.append(curr_goi['name'][j])\n",
    "                count = count + 1\n",
    "           \n",
    "    genes_in_vicinity.append(', '.join(curr_gene_vicinity))\n",
    "    gene_den.append(count)\n",
    "    \n",
    "grna_list['gene_density'] = gene_den\n",
    "grna_list['genes_in_vicinity'] = genes_in_vicinity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Intergenic Region Size [up is after grna and low is before grna irrespective of the strand]\n",
    "grna_to_select = []\n",
    "intg_size = []\n",
    "gene_distance_in_prox = []\n",
    "T_P = []\n",
    "genes_3prime_5prime = []\n",
    "\n",
    "for i in range(np.shape(grna_list)[0]):\n",
    "    one_gene_flag = 0\n",
    "    grna_sca = grna_list['scaffold'][i]\n",
    "    grna_loc = grna_list['loc'][i]\n",
    "    grna_str = grna_list['strand'][i]\n",
    "    if grna_str == '+':\n",
    "        curr_grna_loc = grna_loc\n",
    "    else:\n",
    "        curr_grna_loc = len(seqs_df['sequence'][int(grna_list['scaffold'][i].split('_')[1])-1]) - grna_loc\n",
    "\n",
    "    curr_goi = gene_list.loc[gene_list['chrom'] == grna_sca].reset_index(drop=True)\n",
    "    \n",
    "    if np.shape(curr_goi)[0] == 0:\n",
    "        grna_to_select.append('No')\n",
    "        intg_size.append('-')\n",
    "        gene_distance_in_prox.append(['-','-'])\n",
    "        T_P.append(['-','-'])\n",
    "        genes_3prime_5prime.append(['-','-'])\n",
    "        \n",
    "    else:\n",
    "        low_limit = 0\n",
    "        up_limit = len(seqs_df['sequence'][int(grna_sca.split('_')[1])-1])\n",
    "        low_index = 0\n",
    "        up_index = 0\n",
    "\n",
    "        for j in range(np.shape(curr_goi)[0]):\n",
    "            if curr_goi['txStart'][j] < curr_grna_loc and curr_goi['txEnd'][j] < curr_grna_loc: #txend sufficient\n",
    "                if curr_goi['txEnd'][j] > low_limit:\n",
    "                    low_limit = curr_goi['txEnd'][j]\n",
    "                    low_index = j\n",
    "\n",
    "            if curr_goi['txStart'][j] > curr_grna_loc and curr_goi['txEnd'][j] > curr_grna_loc: #txstart sufficient\n",
    "                if curr_goi['txStart'][j] < up_limit:\n",
    "                    up_limit = curr_goi['txStart'][j]\n",
    "                    up_index = j\n",
    "            \n",
    "        if up_limit == len(seqs_df['sequence'][int(grna_sca.split('_')[1])-1]) or low_limit == 0:\n",
    "            one_gene_flag = 1\n",
    "            \n",
    "            intg_size.append(-1)\n",
    "            if low_limit == 0:\n",
    "                if grna_str == '+':\n",
    "                    gene_distance_in_prox.append([np.abs(up_limit-curr_grna_loc), '-'])\n",
    "                    genes_3prime_5prime.append([curr_goi['name'][up_index],'-'])\n",
    "                else:\n",
    "                    gene_distance_in_prox.append(['-', np.abs(up_limit-curr_grna_loc)])\n",
    "                    genes_3prime_5prime.append(['-',curr_goi['name'][up_index]])\n",
    "                    \n",
    "                if curr_goi['strand'][up_index] == '+':\n",
    "                    if grna_str == '+':\n",
    "                        T_P.append(['P','-'])\n",
    "                    else:\n",
    "                        T_P.append(['-','P'])\n",
    "                        \n",
    "                    if up_limit - curr_grna_loc > 425:\n",
    "                        grna_to_select.append('Yes')\n",
    "                    else:\n",
    "                        grna_to_select.append('No')\n",
    "                else:\n",
    "                    if grna_str == '+':\n",
    "                        T_P.append(['T','-'])\n",
    "                    else:\n",
    "                        T_P.append(['-','T'])\n",
    "                        \n",
    "                    if up_limit - curr_grna_loc > 125:\n",
    "                        grna_to_select.append('Yes')\n",
    "                    else:\n",
    "                        grna_to_select.append('No')\n",
    "                        \n",
    "            else:\n",
    "                if grna_str == '+':\n",
    "                    gene_distance_in_prox.append(['-', np.abs(low_limit-curr_grna_loc)])\n",
    "                    genes_3prime_5prime.append(['-',curr_goi['name'][low_index]])\n",
    "                else:\n",
    "                    gene_distance_in_prox.append([np.abs(low_limit-curr_grna_loc), '-'])\n",
    "                    genes_3prime_5prime.append([curr_goi['name'][low_index],'-'])\n",
    "                    \n",
    "                if curr_goi['strand'][low_index] == '+':\n",
    "                    if grna_str == '+':\n",
    "                        T_P.append(['-','T'])\n",
    "                    else:\n",
    "                        T_P.append(['T','-'])\n",
    "                        \n",
    "                    if curr_grna_loc - low_limit > 125:\n",
    "                        grna_to_select.append('Yes')\n",
    "                    else:\n",
    "                        grna_to_select.append('No')\n",
    "                        \n",
    "                else:\n",
    "                    if grna_str == '+':\n",
    "                        T_P.append(['-','P'])\n",
    "                    else:\n",
    "                        T_P.append(['P','-']) \n",
    "                            \n",
    "                    if curr_grna_loc - low_limit > 425:\n",
    "                        grna_to_select.append('Yes')\n",
    "                    else:\n",
    "                        grna_to_select.append('No')\n",
    "                        \n",
    "        else:\n",
    "            intg_size.append(up_limit - low_limit)\n",
    "            \n",
    "            if grna_str == '+':\n",
    "                gene_distance_in_prox.append([np.abs(up_limit-curr_grna_loc), np.abs(low_limit-curr_grna_loc)])\n",
    "                genes_3prime_5prime.append([curr_goi['name'][up_index],curr_goi['name'][low_index]])\n",
    "            else:\n",
    "                gene_distance_in_prox.append([np.abs(low_limit-curr_grna_loc), np.abs(up_limit-curr_grna_loc)])\n",
    "                genes_3prime_5prime.append([curr_goi['name'][low_index],curr_goi['name'][up_index]])\n",
    "                \n",
    "            if curr_goi['strand'][low_index] == '+' and curr_goi['strand'][up_index] == '+':\n",
    "                if grna_str == '+':\n",
    "                    T_P.append(['P','T'])\n",
    "                else:\n",
    "                    T_P.append(['T','P'])\n",
    "                    \n",
    "                if up_limit - curr_grna_loc > 425 and curr_grna_loc - low_limit > 125:\n",
    "                    grna_to_select.append('Yes')\n",
    "                else:\n",
    "                    grna_to_select.append('No')\n",
    "\n",
    "            elif curr_goi['strand'][low_index] == '+' and curr_goi['strand'][up_index] == '-':\n",
    "                T_P.append(['T','T'])\n",
    "                if up_limit - curr_grna_loc > 125 and curr_grna_loc - low_limit > 125:\n",
    "                    grna_to_select.append('Yes')\n",
    "                else:\n",
    "                    grna_to_select.append('No')\n",
    "\n",
    "            elif curr_goi['strand'][low_index] == '-' and curr_goi['strand'][up_index] == '+':\n",
    "                T_P.append(['P','P'])\n",
    "                if up_limit - curr_grna_loc > 425 and curr_grna_loc - low_limit > 425:\n",
    "                    grna_to_select.append('Yes')\n",
    "                else:\n",
    "                    grna_to_select.append('No')\n",
    "\n",
    "            else:\n",
    "                if grna_str == '+':\n",
    "                    T_P.append(['T','P'])\n",
    "                else:\n",
    "                    T_P.append(['P','T'])\n",
    "                    \n",
    "                if up_limit - curr_grna_loc > 125 and curr_grna_loc - low_limit > 425:\n",
    "                    grna_to_select.append('Yes')\n",
    "                else:\n",
    "                    grna_to_select.append('No')\n",
    "\n",
    "grna_list['intergenic_region_size'] = intg_size\n",
    "grna_list['3\"_gene_dist'] = [i[0] for i in gene_distance_in_prox]\n",
    "grna_list['3\"_T_or_P'] = [i[0] for i in T_P]\n",
    "grna_list['3\"_gene'] = [i[0] for i in genes_3prime_5prime]\n",
    "grna_list['5\"_gene_dist'] = [i[1] for i in gene_distance_in_prox]\n",
    "grna_list['5\"_T_or_P'] = [i[1] for i in T_P]\n",
    "grna_list['5\"_gene'] = [i[1] for i in genes_3prime_5prime]\n",
    "grna_list['grna_to_select'] = grna_to_select\n",
    "\n",
    "grna_list_f = grna_list[grna_list.grna_to_select == 'Yes'].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "939\n"
     ]
    }
   ],
   "source": [
    "#Gene Essentiality Information from Saccharomyces Homology Comparison\n",
    "homology_data = pd.read_csv('blast.csv', delimiter='\\t', header=None)  \n",
    "\n",
    "essential_protein_id = []\n",
    "for i in range(np.shape(homology_data)[0]):\n",
    "    essential_protein_id.append(int(homology_data[1][i].split('|')[2]))\n",
    "    \n",
    "unique_essential_protein_id = list(set(essential_protein_id))\n",
    "print(len(unique_essential_protein_id))\n",
    "\n",
    "mrna_name = []\n",
    "for i in range(len(unique_essential_protein_id)):\n",
    "    mrna_name.append(name_conversion['mrna'][name_conversion[name_conversion['protein ID'] == unique_essential_protein_id[i]].index.tolist()[0]])\n",
    "\n",
    "gene_ess_count = []\n",
    "for i in range(np.shape(grna_list_f)[0]):\n",
    "    curr_gene_ess_count = 0\n",
    "    curr_gene_vicinity_2 = [grna_list_f['5\"_gene'][i],grna_list_f['3\"_gene'][i]]\n",
    "    for j in range(len(curr_gene_vicinity_2)):\n",
    "        if curr_gene_vicinity_2[j] in mrna_name:\n",
    "            curr_gene_ess_count = curr_gene_ess_count + 1\n",
    "            \n",
    "    if curr_gene_ess_count > 0:\n",
    "        gene_ess_count.append(curr_gene_ess_count)\n",
    "    else:\n",
    "        gene_ess_count.append('NA')\n",
    "        \n",
    "grna_list_f['adjacent_gene_ess'] = gene_ess_count"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Output\n",
    "grna_list_filtered_f = grna_list_f.loc[(grna_list_f['gene_density'] > 4) & (grna_list_f['intergenic_region_size'] < 3000) & (grna_list_f['intergenic_region_size'] > 0)].reset_index(drop=True)\n",
    "grna_list_filtered_f['intergenic_region_size'] = grna_list_filtered_f['intergenic_region_size'].replace([-1], '-')\n",
    "#grna_list_filtered_f.to_csv('grna_list_with_gene_vicinity_info_filtered.csv',index=False)\n",
    "\n",
    "grna_list_f['intergenic_region_size'] = grna_list_f['intergenic_region_size'].replace([-1], '-')\n",
    "del grna_list_f['grna_to_select']\n",
    "#grna_list_f.to_csv('grna_list_with_gene_vicinity_info_total.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Transcriptomics data\n",
    "expr_data = pd.read_excel('Top10_based_on_TPM.xlsx')\n",
    "\n",
    "mrna_name = []\n",
    "\n",
    "for i in range(np.shape(expr_data)[0]):\n",
    "    mrna_name.append(name_conversion['mrna'][name_conversion[name_conversion['protein ID'] == int(expr_data['Conserved stable gene'][i].split('Issorie2|')[1])].index.tolist()[0]])\n",
    "\n",
    "expr_data['mrna'] = mrna_name\n",
    "\n",
    "avg_expr_vic_2 = []\n",
    "\n",
    "for i in range(np.shape(grna_list_f)[0]):\n",
    "    if grna_list_f['gene_density'][i] > 0:\n",
    "        curr_expr_vic_2 = []\n",
    "            \n",
    "        curr_gene_vicinity_2 = [grna_list_f['5\"_gene'][i],grna_list_f['3\"_gene'][i]]\n",
    "        for j in range(len(curr_gene_vicinity_2)):\n",
    "            if curr_gene_vicinity_2[j] in mrna_name:\n",
    "                curr_index = mrna_name.index(curr_gene_vicinity_2[j])\n",
    "                curr_expr_vic_2.append(np.median([expr_data['IO126_LOD_16h'][curr_index], expr_data['IO126_LOD_16h.1'][curr_index], expr_data['IO126_LOD_16h.2'][curr_index]]))\n",
    "        \n",
    "        if len(curr_expr_vic_2) > 0:\n",
    "            avg_expr_vic_2.append(np.mean(curr_expr_vic_2))\n",
    "        else:\n",
    "            avg_expr_vic_2.append('NA')\n",
    "        \n",
    "    else:\n",
    "        avg_expr_vic_2.append('NA')\n",
    "        \n",
    "grna_list_f['avg_expr_vic_2'] = avg_expr_vic_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "#On-target activity (RULE SET 2, Doench 2016)\n",
    "on_target_file = 'on_target_activity.txt'\n",
    "with open(on_target_file) as file:\n",
    "    lines = file.readlines()\n",
    "    on_target_activity = [np.float64(line.rstrip().split(': ')[1]) for line in lines]\n",
    "    \n",
    "grna_list_f['on_target_activity'] = on_target_activity"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Output\n",
    "grna_list_f.to_csv('output.csv',index=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.11"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
